### ISSUE TITLE
Error when attempting to bind to component properties initialized outside constructor

### ISSUE THREAD

**Issue by theodorejb** [angular/angular#19587]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request
</code></pre>

## Current behavior
Angular 5.0.0-rc.1 throws "TypeError: e.component.newUser is undefined" when running the following component (after AoT compilation):

```typescript
import {Component} from '@angular/core';

@Component({
    template: `
        <form>
            <input name="name" [(ngModel)]="newUser.fullName"/>
        </form>`,
})
export class RegisterComponent {
    public newUser = {
        fullName: ''
    };
}
```

## Workaround
It works if I move the newUser assignment into a constructor method, but this wasn't necessary in Angular 4.x.

## Expected behavior
No error should be thrown. `newUser` properties should be bound to form fields as expected when the component is initialized.

## What is the motivation / use case for changing the behavior?
The above component worked without errors in Angular 4.x.

## Environment

<pre><code>
Angular version: 5.0.0-rc.1

Browser:
- [x] Chrome (desktop) version 61
- [x] Firefox version 56
- [x] Edge version 15
 
For Tooling issues:
- Node version: 8.6
- Platform:  Windows
</code></pre>

**Comment by petebacondarwin** [comment]
I think this is a bug in ngrx.

The workaround is to add `"angularCompilerOptions": { "annotationsAs": "decorators" }` to your tsconfig to disable tsickle.

**Comment by petebacondarwin** [comment]
@tbosch  ^^^

**Comment by tbosch** [comment]
Pete, where does "e.component" come from, as it's not present in the example?

**Comment by petebacondarwin** [comment]
I am only extrapolating from the problem that I had with the aot-compiler example, which @tbosch kindly helped me with. In that case it seems that if you don't set this configuration property the `component.factory.js` file that is created by `ngc` strips out fields that are initialized outside of the constructor.

So in this case the `e.component` is how the `RegisterComponent` instance is being referenced in the factory file.

@theodorejb could confirm this by publishing the code of the factory file.

**Comment by theodorejb** [comment]
I'm away from my home PC right now, but I compiled the component on a different PC and the error in Firefox changed to "TypeError: _co.newUser is undefined". In Chrome the error is "TypeError: Cannot read property 'fullName' of undefined". Here's the generated register.ngfactory.js file:
[register.ngfactory.js.txt](https://github.com/angular/angular/files/1362932/register.ngfactory.js.txt)

**Comment by theodorejb** [comment]
Adding `"angularCompilerOptions": { "annotationsAs": "decorators" }` to my tsconfig did not fix the error.

**Comment by petebacondarwin** [comment]
Sorry to hear that :-(

**Comment by petebacondarwin** [comment]
Oops sorry, I meant to say it was the actually compiled `RegisterComponent`, not the factory that I had a problem with.

**Comment by tbosch** [comment]
@theodorejb Sorry, but we need a better reproduction for this. Could you create a branch on Github that shows the problem? Could you check that you don't assign `fullUser` a new value somewhere else? The factory looks fine, it just updates the property inside of `fullUser`, but not the `fullUser` object itself...

Could you also paste the compiled `register.js` file that contains the component class?

**Comment by sarunint** [comment]
Hi, I was able to reproduce this issue. The difference between mine and @theodorejb is that `"angularCompilerOptions": { "annotationsAs": "decorators" }` was able to fix this problem.

Here is my repro: https://github.com/sarunint/angular-scratch

Step to reproduce: run `ngc -p ./src/tsconfig.app.aot.json`

This is the uncompiled `app.component.ts`:

```typescript
import { Component, OnInit } from '@angular/core';

@Component({
  selector: 'app-root',
  template: `<div>{{text}}</div><input type="text" [(ngModel)]="text">`
})
export class AppComponent implements OnInit {
  text = 'Hello';

  constructor() { }

  ngOnInit() { }
}
```

This is the compiled `app.component.js`:

```javascript
import { Component } from '@angular/core';
export class AppComponent {
    constructor() { }
    ngOnInit() { }
}
//# sourceMappingURL=app.component.js.map
```

There should be `this.text = 'Hello'` in the constructor of the compiled component.

**Comment by theodorejb** [comment]
@tbosch I'm still getting this error in Angular 5.0.0-rc.2. Here's the contents of the compiled register.js file:

```javascript
import { Component } from '@angular/core';
var RegisterComponent = /** @class */ (function () {
    function RegisterComponent() {
    }
    return RegisterComponent;
}());
export { RegisterComponent };
```

**Comment by sarunint** [comment]
Hey @theodorejb, are you using TypeScirpt 2.5? Because I tried downgrade it to 2.4 and there is no issue.

**Comment by theodorejb** [comment]
@sarunint Wow, it does work after I downgraded to TypeScript 2.4! It's only broken in TypeScript 2.5 and 2.6 RC. Here's the register.js file after switching to TS 2.4:

```javascript
import { Component } from '@angular/core';
var RegisterComponent = (function () {
    function RegisterComponent() {
        this.newUser = {
            fullName: 'my name'
        };
    }
    RegisterComponent.decorators = [
        { type: Component, args: [{
                    template: "\n        <form>\n            <input name=\"name\" [(ngModel)]=\"newUser.fullName\"/>\n        </form>",
                },] },
    ];
    /** @nocollapse */
    RegisterComponent.ctorParameters = function () { return []; };
    return RegisterComponent;
}());
export { RegisterComponent };
```

**Comment by sarunint** [comment]
Well, I'm still surprised that by using TypeScript 2.5, it removes the decorators data, which is not needed while doing AoT compilation anyway. (I may be wrong about this.)

**Comment by tbosch** [comment]
Internally at Google, we are still using TS 2.4, and therefore haven't tested ngc with TS 2.5 extensibly. But it's one of our next tasks and will ship probably with Angular 5.1. 

For now we decided to let users know that we only work with TS 2.4, so that there are no surprises like this via the entry in package.json (see https://github.com/angular/angular/commit/b922743f6e023a4fd76bf9b8e571e727073a1793).

**Comment by xnnkmd** [comment]
@tbosch I noticed that angular 5 does a lot of work to improve static analysis to catch errors early on where they are cheap to fix (i.e. fullTemplateTypeCheck etc.). On the other hand you use old versions of typescript that does not find all the errors staticly that the newest version does. Would it not make sense to support newer versions of typescript more aggresively to get all static checks working and save costs ?

**Comment by sarunint** [comment]
The team is working hardly to make Angular compatible with newer version of TypeScript. (There are some PRs in progress.) But sometimes they are forced to locked down since it will introduce breaking changes. This happened with version 2, in which TypeScript is locked to 2.0.

**Comment by chuckjaz** [comment]
@xnnkmd As @sarunint said, it is a breaking change to change the minimum version of TypeScript we support and the only one we were ready at the beginning of the RC phase was 2.4 (2.5 shipped just after we started) and we try very hard not to make breaking changes once we enter RC. 

The reason we cannot arbitrarily upgrade to a new version of TypeScript when it releases is because TypeScript does not follow semver and they will make breaking changes in minor releases some many of which break us. In the case of 2.4 to 2.5, they changed some of the internal APIs we used to implement Tsickle (which performs the decorator down-leveling). We now have a Tsickle that supports 2.5 but not one that supports 2.4 and 2.5, but we will shortly. When that is available we will increase the supported version of TypeScript to 2.5 (that is, supporting both 2.4 and 2.5). As 2.6 has now been released, we are evaluating when we can support that as well.

**Comment by mkp05** [comment]
The problem appears even in not-annotated classes, but also for example for static fields.

For example transpiling file with simple class:
```typescript
export class Test {
	static readonly something = "Test";
}
```

Produces:
```javascript
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Test = /** @class */ (function () {
    function Test() {
    }
    return Test;
}());
exports.Test = Test;
//# sourceMappingURL=test.js.map
```

Instead of correct:
```javascript
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Test = (function () {
    function Test() {
    }
    Test.something = "Test";
    return Test;
}());
exports.Test = Test;
//# sourceMappingURL=test.js.map
```

**Comment by sarunint** [comment]
As stated above, Angular v5 does not support TypeScript 2.5 and above yet. TypeScript 2.5 support is likely to land in v5.1.

**Comment by mkp05** [comment]
Yes, just wanted give another test-case for tests when working on supporting higher versions of TS... Sorry for that.
